src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs

```diff
@@ -56,149 +67,301 @@ public sealed class RichTextEditorPastedImages
+        }
+
+        return htmlDoc.DocumentNode.OuterHtml;
+    }
+
+    /// <summary>
+    ///     Used by the RTE (and grid RTE) for drag/drop/persisting images.
+    /// </summary>
+    /// <param name="html">HTML from the Rich Text Editor property editor.</param>
+    /// <param name="mediaParentFolder"></param>
+    /// <param name="userId"></param>
+    /// <param name="imageUrlGenerator"></param>
+    /// <returns>Formatted HTML.</returns>
+    [Obsolete("It is not needed to supply the imageUrlGenerator parameter")]
+    public string FindAndPersistPastedTempImages(string html, Guid mediaParentFolder, int userId, IImageUrlGenerator imageUrlGenerator) =>
+        FindAndPersistPastedTempImages(html, mediaParentFolder, userId);
 
-            // Get the new persisted image URL
-            _umbracoContextAccessor.TryGetUmbracoContext(out IUmbracoContext? umbracoContext);
-            IPublishedContent? mediaTyped = umbracoContext?.Media?.GetById(udi.Guid);
-            if (mediaTyped == null)
+    /// <summary>
+    ///     Used by the RTE (and grid RTE) for drag/drop/persisting images.
+    /// </summary>
+    /// <param name="html">HTML from the Rich Text Editor property editor.</param>
+    /// <param name="mediaParentFolder"></param>
+    /// <param name="userId"></param>
+    /// <returns>Formatted HTML.</returns>
+    public string FindAndPersistPastedTempImages(string html, Guid mediaParentFolder, int userId)
+    {
+        // Find all img's that has data-tmpimg attribute
+        // Use HTML Agility Pack - https://html-agility-pack.net
+        var htmlDoc = new HtmlDocument();
+        htmlDoc.LoadHtml(html);
+
+        HtmlNodeCollection? tmpImages = htmlDoc.DocumentNode.SelectNodes($"//img[@{TemporaryImageDataAttribute}]");
+        if (tmpImages == null || tmpImages.Count == 0)
+        {
+            return html;
+        }
+
+        foreach (HtmlNode? img in tmpImages)
+        {
+            // The data attribute contains the path to the tmp img to persist as a media item
+            var tmpImgPath = img.GetAttributeValue(TemporaryImageDataAttribute, string.Empty);
+
+            if (string.IsNullOrEmpty(tmpImgPath))
             {
-                throw new PanicException(
-                    $"Could not find media by id {udi.Guid} or there was no UmbracoContext available.");
+                continue;
             }
 
-            var location = mediaTyped.Url(_publishedUrlProvider);
+            var qualifiedTmpImgPath = _hostingEnvironment.MapPathContentRoot(tmpImgPath);
+
+            PersistMediaItem(mediaParentFolder, userId, img, qualifiedTmpImgPath);
+        }
+
+        return htmlDoc.DocumentNode.OuterHtml;
+    }
+
+    private void PersistMediaItem(Guid mediaParentFolder, int userId, HtmlNode img, string qualifiedTmpImgPath)
+    {
+        var absoluteTempImagePath = Path.GetFullPath(qualifiedTmpImgPath);
+
+        if (IsValidPath(absoluteTempImagePath) == false)
```

Comment:
Lets change this to pattern matching while we're at it 🔥
```suggestion
        if (IsValidPath(absoluteTempImagePath) is false)
```